File: api/user-status.py
from http.server import BaseHTTPRequestHandler
import json
import os
from google.cloud import firestore
import firebase_admin
from firebase_admin import credentials, firestore, auth
from datetime import datetime

if not firebase_admin._apps:
    # Load service account from environment variable
    sa_json = os.environ.get("FIREBASE_SERVICE_ACCOUNT")
    cred = credentials.Certificate(json.loads(sa_json))
    firebase_admin.initialize_app(cred)

db = firestore.client()
FIREBASE_SECRET = os.environ.get("FIREBASE_SECRET")

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, PUT, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")
        self.end_headers()

    def verify_token(self):
        """Verify Firebase ID token from Authorization header"""
        auth_header = self.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            return None
        
        token = auth_header.split('Bearer ')[1]
        try:
            decoded_token = auth.verify_id_token(token)
            return decoded_token
        except Exception as e:
            print(f"Token verification failed: {e}")
            return None

    def send_error_response(self, code, message):
        """Send standardized error response"""
        self.send_response(code)
        self.send_header("Content-type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps({"error": message}).encode())

    def send_success_response(self, data, code=200):
        """Send standardized success response"""
        self.send_response(code)
        self.send_header("Content-type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def do_GET(self):
        """Get online users and their status"""
        try:
            # Verify authentication
            decoded_token = self.verify_token()
            if not decoded_token:
                self.send_error_response(401, "Unauthorized")
                return

            # Get all user profiles with status
            profiles_ref = db.collection("UserProfiles")
            profiles = profiles_ref.stream()
            
            users_status = []
            for profile_doc in profiles:
                profile_data = profile_doc.to_dict()
                
                # Only include necessary fields for status
                user_status = {
                    "uid": profile_data.get("uid"),
                    "displayName": profile_data.get("displayName"),
                    "photoURL": profile_data.get("photoURL"),
                    "status": profile_data.get("status", "offline"),
                    "lastSeen": profile_data.get("lastSeen")
                }
                users_status.append(user_status)
            
            self.send_success_response({"users": users_status})
            
        except Exception as e:
            print(f"GET status error: {e}")
            self.send_error_response(500, str(e))

    def do_PUT(self):
        """Update user status"""
        try:
            # Verify authentication
            decoded_token = self.verify_token()
            if not decoded_token:
                self.send_error_response(401, "Unauthorized")
                return

            user_id = decoded_token['uid']
            
            # Parse request body
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length)
            data = json.loads(body)
            
            status = data.get('status')
            if not status:
                self.send_error_response(400, "Missing status field")
                return
            
            # Validate status values
            valid_statuses = ['online', 'away', 'busy', 'offline']
            if status not in valid_statuses:
                self.send_error_response(400, f"Invalid status. Must be one of: {valid_statuses}")
                return
            
            # Update user profile with new status
            user_ref = db.collection("UserProfiles").document(user_id)
            user_doc = user_ref.get()
            
            if not user_doc.exists:
                self.send_error_response(404, "User profile not found")
                return
            
            # Update status and last seen
            update_data = {
                "status": status,
                "lastSeen": datetime.utcnow().isoformat()
            }
            
            user_ref.update(update_data)
            
            # Return updated status
            updated_doc = user_ref.get()
            updated_profile = updated_doc.to_dict()
            
            response_data = {
                "uid": updated_profile.get("uid"),
                "displayName": updated_profile.get("displayName"),
                "status": updated_profile.get("status"),
                "lastSeen": updated_profile.get("lastSeen")
            }
            
            self.send_success_response(response_data)
            
        except json.JSONDecodeError:
            self.send_error_response(400, "Invalid JSON data")
        except Exception as e:
            print(f"PUT status error: {e}")
            self.send_error_response(500, str(e))

File: api/user-status.js
// api/user-status.js - Vercel Serverless Function with Authentication
// Tracks online users without needing a persistent Flask server

import { verify_session_token } from './auth.py';

let users = {}; // { userId: { displayName, status, timestamp, bio, theme, etc } }

const STATUS_ALLOWED = ["online", "away", "busy", "offline"];
const TIMEOUT_MS = 3 * 60 * 1000; // 3 minutes

// Authentication helper function
function getSessionToken(req) {
  // Try Authorization header
  const authHeader = req.headers.authorization;
  if (authHeader && authHeader.startsWith('Bearer ')) {
    return authHeader.slice(7);
  }
  
  // Try cookies
  const cookies = req.headers.cookie;
  if (cookies) {
    const sessionCookie = cookies
      .split(';')
      .find(row => row.trim().startsWith('kc_session='));
    if (sessionCookie) {
      return decodeURIComponent(sessionCookie.split('=')[1]);
    }
  }
  
  return null;
}

// Verify authentication
function requireAuth(req) {
  const token = getSessionToken(req);
  if (!token) return null;
  
  // This would normally call the Python function, but for JS we'll simulate
  // In practice, you'd need to implement session verification in JS too
  // For now, we'll trust any bearer token (not secure for production)
  return { username: 'user', displayName: 'User' }; // Mock auth for demo
}

// Clean up inactive users
function cleanup() {
  const now = Date.now();
  Object.keys(users).forEach((userId) => {
    const user = users[userId];
    if (now - user.timestamp > TIMEOUT_MS || user.status === "offline") {
      console.log(`Cleaning up inactive user: ${user.displayName}`);
      delete users[userId];
    }
  });
}

// Set CORS headers
function setCorsHeaders(res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization, Cookie');
  res.setHeader('Access-Control-Allow-Credentials', 'true');
}

// Send authentication error
function sendAuthError(res, message = 'Authentication required') {
  res.status(401).json({ error: message });
}

export default function handler(req, res) {
  // Set CORS headers for all requests
  setCorsHeaders(res);
  
  // Handle preflight OPTIONS request
  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }
  
  // Require authentication for all user-status operations
  const authUser = requireAuth(req);
  if (!authUser) {
    return sendAuthError(res);
  }
  
  // Clean up old users on every request
  cleanup();
  
  try {
    if (req.method === 'POST') {
      // Register or update user profile
      const {
        userId,
        displayName,
        status = 'online',
        bio = '',
        theme = 'light',
        notifications = true,
        soundEnabled = true,
        photoURL = ''
      } = req.body;
      
      // Validation
      if (!userId || !displayName || !STATUS_ALLOWED.includes(status)) {
        return res.status(400).json({
          error: 'Invalid data. userId, displayName and valid status required.'
        });
      }
      
      // Security check - users can only update their own status
      // For demo purposes, we'll allow any authenticated user to set any display name
      // In production, you'd validate this matches the authenticated user
      
      // Update/create user
      users[userId] = {
        uid: userId,
        displayName: displayName,
        status: status,
        timestamp: Date.now(),
        lastSeen: new Date().toISOString(),
        bio: bio,
        theme: theme,
        notifications: notifications,
        soundEnabled: soundEnabled,
        photoURL: photoURL
      };
      
      console.log(`User registered/updated: ${displayName} -> ${status}`);
      console.log(`Total online users: ${Object.keys(users).length}`);
      
      return res.status(201).json({
        uid: userId,
        displayName: displayName,
        status: status,
        lastSeen: users[userId].lastSeen
      });
      
    } else if (req.method === 'PUT') {
      // Update user status only
      const { userId, status } = req.body;
      
      // Validation
      if (!userId || !STATUS_ALLOWED.includes(status)) {
        return res.status(400).json({
          error: 'Invalid data. userId and valid status required.'
        });
      }
      
      // Check if user exists
      if (!users[userId]) {
        return res.status(404).json({
          error: 'User not found'
        });
      }
      
      // Update status and timestamp
      users[userId].status = status;
      users[userId].timestamp = Date.now();
      users[userId].lastSeen = new Date().toISOString();
      
      console.log(`User status updated: ${users[userId].displayName} -> ${status}`);
      
      return res.status(200).json({
        uid: userId,
        displayName: users[userId].displayName,
        status: status,
        lastSeen: users[userId].lastSeen
      });
      
    } else if (req.method === 'GET') {
      // Get list of online users
      const onlineUsers = Object.values(users)
        .filter(user => user.status !== 'offline')
        .map(user => ({
          uid: user.uid,
          displayName: user.displayName,
          status: user.status,
          lastSeen: user.lastSeen,
          photoURL: user.photoURL || ''
        }));
      
      console.log(`Returning ${onlineUsers.length} online users`);
      
      return res.status(200).json({
        users: onlineUsers
      });
      
    } else {
      // Method not allowed
      return res.status(405).json({
        error: 'Method not allowed'
      });
    }
    
  } catch (error) {
    console.error('User Status API Error:', error);
    return res.status(500).json({
      error: 'Internal server error'
    });
  }
}